### ISSUE TITLE
no-dso was not useless

### ISSUE THREAD

**Issue by benjaminp** [openssl/openssl#9036]
5fba3afad01707f4a8856a35500de007a8a256ec, which first appeared in 1.1.1c, made the `no-dso` Configure option completely a noop on the grounds that it failed to define `OPENSSL_NO_DSO`. However, `no-dso` was still useful because it resulted in a empty `crypto/include/internal/dso_conf.h` header, which meant OpenSSL did not depend on `libdl.so`. Now in OpenSSL 1.1.1c, there doesn't seem to be a good way to force dso loading off on platforms that support it. 

Could the 1.1.1b behavior of `no-dso` be restored?

**Comment by levitte** [comment]
I have no idea why you think you avoided libdl with no-dso.  I just tried the experiment myself by checking out and building `OpenSSL_1_1_1b`, and 'lo and behold:

```
: ; grep ^VERSION= Makefile
VERSION=1.1.1b
: ; ./configdata.pm -c

Command line (with current working directory = .):

    /usr/bin/perl ../1.1.1/Configure linux-x86_64 no-dso

Perl information:

    /usr/bin/perl
    5.28.1 for x86_64-linux-gnu-thread-multi
: ; ./util/shlib_wrap.sh ldd apps/openssl
	linux-vdso.so.1 (0x00007fff197de000)
	libssl.so.1.1 => ./util/../libssl.so.1.1 (0x00007fd8fcb33000)
	libcrypto.so.1.1 => ./util/../libcrypto.so.1.1 (0x00007fd8fc84d000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fd8fc7f1000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fd8fc7d0000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fd8fc60f000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fd8fcc85000)
: levitte@lapdog:~/gitwrk/openssl.net/official/_build-1.1.1
: ; ./util/shlib_wrap.sh ldd libcrypto.so
	linux-vdso.so.1 (0x00007ffef1fd6000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f104f686000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f104f665000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f104f4a4000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f104f9ca000)
```

**Comment by benjaminp** [comment]
Sure, `-ldl` might still be in `LDFLAGS`, but the dependency isn't written if `-as-needed` is in effect.

**Comment by levitte** [comment]
Are you sure about that?

```
: ; ./configdata.pm -c

Command line (with current working directory = .):

    /usr/bin/perl ../1.1.1/Configure linux-x86_64 no-dso -Wl,--as-needed

Perl information:

    /usr/bin/perl
    5.28.1 for x86_64-linux-gnu-thread-multi
: ; LANG=C objdump -x libcrypto.so|grep libdl
  NEEDED               libdl.so.2
  required from libdl.so.2:
```

**Comment by benjaminp** [comment]
Presumably you have to insert `-as-needed` before `-ldl` in the linker command line. My system doesn't need to manipulate `LDFLAGS` at all because Ubuntu has patched the GCC driver to always pass `-as-needed` to the linker at the beginning.

Anyway, I don't want to become too lost in the weeds of linker command lines here. The old `no-dso` behavior was certainly fragile. My hope is that it can fixed up (e.g., by actually omitting `-ldl`) rather become a complete no-op.

**Comment by levitte** [comment]
`-Wl,--as-needed` ends up in `LIB_LDFLAGS`, which is used early in the command line, while `-ldl` ends up in `LIB_EX_LIBS`, which is in the very last bit of the linking command line.

So uhm, I still do not understand how you avoided linking with libdl. How about you tell me?

**Comment by levitte** [comment]
BTW, if you really want to be rid of DSO capabilities, there *is* a way; create your own config target, based on the one you normally use.  So let's say you normally use `linux-x86_64`, then you can put this in a file, say `whatever.conf`:

``` perl
(
    'my-linux-x86_64' => {
        inherit_from    => [ 'linux-x86_64' ],
        dso_scheme     => undef,
    }
);
```

Then you can use that like this:

```
$ ./Configure --config whatever.conf my-linux-x86_64
```

**Comment by benjaminp** [comment]
Hopefully, this session is reproducible almost anywhere:
```
$ docker run -it ubuntu:18.04
# apt install curl build-essential
# curl https://www.openssl.org/source/openssl-1.1.1b.tar.gz | tar xz
# cd openssl-1.1.1b

# ./Configure linux-x86_64 no-dso
# make -s
# readelf -d libcrypto.so.1.1 | grep NEEDED
 0x0000000000000001 (NEEDED)             Shared library: [libpthread.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]

# make distclean
# ./Configure linux-x86_64
# make -s
# readelf -d libcrypto.so.1.1 | grep NEEDED
 0x0000000000000001 (NEEDED)             Shared library: [libdl.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libpthread.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]

# cd ..
# curl https://www.openssl.org/source/openssl-1.1.1c.tar.gz | tar xz
# cd openssl-1.1.1c
# ./Configure linux-x86_64 no-dso
# make -s
# readelf -d libcrypto.so.1.1 | grep NEEDED
 0x0000000000000001 (NEEDED)             Shared library: [libdl.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libpthread.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
```

**Comment by TyrasNistru** [comment]
Aside from the need to link to -libdl, the -no-dso has other uses. There are platforms where dynamic library loading does not work properly or has specific requirements. 

WASM/Emscripten (my case) is an example: to use dload, both the loading code and the library being loaded has to be compiled in a special way, and, for this platform, AFAIK, the DSO brings no advantage.

**Comment by InspectorConstructor** [comment]
As of at least openssl 1.1.1d, if you wish to specify a config file to Configure (to forcibly rid DSO from the build as described above,) the command is `$ ./Configure --config**=**whatever.conf my-linux-x86_64` (note the '=' character.)

**Comment by InspectorConstructor** [comment]
> BTW, if you really want to be rid of DSO capabilities, there _is_ a way; create your own config target, based on the one you normally use. So let's say you normally use `linux-x86_64`, then you can put this in a file, say `whatever.conf`:
> 
> ```perl
> (
>     'my-linux-x86_64' => {
>         inherit_from    => [ 'linux-x86_64' ],
>         dso_scheme     => undef,
>     }
> );
> ```
> 
> Then you can use that like this:
> 
> ```
> $ ./Configure --config whatever.conf my-linux-x86_64
> ```

As of at least openssl 1.1.1d, the flag for specifying a config file is `--config=whatever.conf` (note the '=' character.) If you use a space instead of an '=', Configure gives an extremely unhelpful error.

**Comment by mattcaswell** [comment]
This was fixed by #9889. Closing.
